apps/ui/src/components/BottomAppBar.tsx

```diff
@@ -0,0 +1,466 @@
+import ExpandMore from "@mui/icons-material/ExpandMore";
+import Tooltip from "@mui/material/Tooltip";
+import IconButton from "@mui/material/IconButton";
+import LinkOffIcon from "@mui/icons-material/LinkOff";
+import Paper from "@mui/material/Paper";
+import PlayArrowIcon from "@mui/icons-material/PlayArrow";
+import Popover from "@mui/material/Popover";
+import StopIcon from "@mui/icons-material/Stop";
+import RefreshIcon from "@mui/icons-material/Refresh";
+import CloudUploadIcon from "@mui/icons-material/CloudUpload";
+import List from "@mui/material/List";
+import ListItem from "@mui/material/ListItem";
+
+import { gql, useMutation } from "@apollo/client";
+import { useStore } from "zustand";
+
+import { RepoContext } from "../lib/store";
+import {
+  ListItemIcon,
+  ListItemButton,
+  Stack,
+  ListItemText,
+} from "@mui/material";
+import { myNanoId } from "../lib/utils/utils";
+import { match } from "ts-pattern";
+
+const RuntimeItem = ({ runtimeId }) => {
+  const store = useContext(RepoContext)!;
+  const runtimeMap =
+    useStore(store, (state) => state.getRuntimeMap()) || new Map();
+  // Observe runtime change
+  const runtimeChanged = useStore(store, (state) => state.runtimeChanged);
+  // A dummy useEffect to indicate that runtimeChanged is used.
+  useEffect(() => {}, [runtimeChanged]);
+  const activeRuntime = useStore(store, (state) => state.activeRuntime);
+  const setActiveRuntime = useStore(store, (state) => state.setActiveRuntime);
+  const runtime = runtimeMap.get(runtimeId) || { wsStatus: "unknown" };
+  const repoId = useStore(store, (state) => state.repoId);
+
+  const [requestKernelStatus] = useMutation(
+    gql`
+      mutation RequestKernelStatus($runtimeId: String) {
+        requestKernelStatus(runtimeId: $runtimeId)
+      }
+    `,
+    { context: { clientName: "spawner" } }
+  );
+  const [interruptKernel] = useMutation(
+    gql`
+      mutation InterruptKernel($runtimeId: String) {
+        interruptKernel(runtimeId: $runtimeId)
+      }
+    `,
+    { context: { clientName: "spawner" } }
+  );
+  const [killRuntime] = useMutation(
+    gql`
+      mutation KillRuntime($runtimeId: String, $repoId: String) {
+        killRuntime(runtimeId: $runtimeId, repoId: $repoId)
+      }
+    `,
+    { context: { clientName: "spawner" } }
+  );
+
+  const [disconnectRuntime] = useMutation(
+    gql`
+      mutation DisconnectRuntime($runtimeId: String, $repoId: String) {
+        disconnectRuntime(runtimeId: $runtimeId, repoId: $repoId)
+      }
+    `,
+    { context: { clientName: "spawner" } }
+  );
+
+  const runtimeStatusColors = { idle: "green", busy: "yellow", default: "red" };
+  const runtimeStatus = runtime.status ?? "default";
+  const wsStatusConfig = {
+    connected: { title: "connected", color: "green" },
+    connecting: { title: "connecting", color: "yellow" },
+    default: { title: "disconnected", color: "red" },
+  };
+  const wsStatus = wsStatusConfig[runtime.wsStatus || "default"];
```

Comment:
[2/3] And if that's `unknown`, this wsStatusConfig doesn't have the key, which leads to `wsStatus===undefined`

Reply:
Adding some type annotations can help you catch such bugs early and more comprehensively.